### PR TITLE
CU-cn24gw CU-cn24dh Remove extension and higherkind annotations for Endo and Eval

### PR DESCRIPTION
**Pull request by danimontoya**
- [x] Remove extension annotation for Endo
- [x] Remove extension and higherkind annotations for Eval
- [x] Deprecate extension for Endo
- [x] Deprecate extension EvalApplicative
- [x] Deprecate extension EvalApply
- [x]  Deprecate extension EvalBimonad
- [x] Deprecate extension EvalComonad
- [x] Deprecate extension EvalFunctor
- [x] Deprecate extension EvalMonad

### CODE REVIEW

**Comment by franciscodr** [comment]
Task linked: [CU-cn24dh Remove extension annotation Endo](https://app.clickup.com/t/cn24dh)

**Comment by franciscodr** [comment]
Task linked: [CU-cn24gw Remove extension annotation Eval](https://app.clickup.com/t/cn24gw)

File: arrow-core/src/main/kotlin/arrow/core/extensions/endo/monoid/EndoMonoid.kt [hunk 1]
```diff
@@ -27,14 +27,14 @@ internal val monoid_singleton: EndoMonoid<Any?> = object : EndoMonoid<Any?> {}
 @Deprecated(
   "@extension kinded projected functions are deprecated",
   ReplaceWith(
-  "combineAll()",
-  "arrow.core.combineAll"
+    "if (isEmpty()) empty() else reduce { a, b -> a.combine(b) }"
```
Comment (nomisRev) [inline arrow-core/src/main/kotlin/arrow/core/extensions/endo/monoid/EndoMonoid.kt:30]: `combine` is not defined on `Endo`.

**Review by nomisRev — CHANGES_REQUESTED** [review]
Looks great! Only `combine` missing on `Endo` itself.

File: arrow-core-data/src/main/kotlin/arrow/core/Endo.kt [hunk 2]
```diff
@@ -9,6 +9,9 @@ data class Endo<A>(val f: (A) -> A) {
   companion object
 }
 
+fun <A> Endo<A>.combine(g: Endo<A>): Endo<A> =
+  Endo(f.compose(g.f))
+
```
Comment (nomisRev) [inline arrow-core-data/src/main/kotlin/arrow/core/Endo.kt]:
```suggestion
  fun combine(g: Endo<A>): Endo<A> =
    Endo(f.compose(g.f))
}
```

File: arrow-core/src/main/kotlin/arrow/core/extensions/endo/monoid/EndoMonoid.kt [hunk 3]
```diff
@@ -46,18 +40,20 @@ fun <A> Collection<Endo<A>>.combineAll(): Endo<A> = arrow.core.Endo.monoid<A>().
...
+fun <A> combineAll(arg0: List<Endo<A>>): Endo<A> =
+  arrow.core.Endo
+    .monoid<A>()
+    .combineAll(arg0) as arrow.core.Endo<A>
 
 @Suppress(
   "UNCHECKED_CAST",
   "NOTHING_TO_INLINE"
 )
+@Deprecated("Monoid typeclasses is deprecated. Use concrete methods on Endo")
```
Comment (nomisRev) [inline arrow-core/src/main/kotlin/arrow/core/extensions/endo/monoid/EndoMonoid.kt:57]:
```suggestion
@Deprecated("Extension projections are deprecated. Use endo on Monoid.", ReplaceWith("Monoid.endo()", "arrow.typeclasses.Monoid", "arrow.core.endo")
```

File: arrow-core-data/src/main/kotlin/arrow/core/Eval.kt [hunk 4]
```diff
@@ -75,6 +87,14 @@ sealed class Eval<out A> : EvalOf<A> {
...
 
+    fun <A, B> tailRecM(a: A, f: (A) -> Eval<Either<A, B>>): Eval<B> =
+      f(a).flatMap { eval: Either<A, B> ->
+        when (eval) {
+          is Either.Left -> tailRecM(eval.a, f)
+          is Either.Right -> just(eval.b)
+        }
+      }
+
     fun <A> just(a: A): Eval<A> =
```
Comment (nomisRev) [inline arrow-core-data/src/main/kotlin/arrow/core/Eval.kt:98]: Should be deprecated in favor of `now`

File: arrow-core/src/main/kotlin/arrow/core/extensions/eval/applicative/EvalApplicative.kt [hunk 5]
```diff
@@ -32,8 +23,8 @@ internal val applicative_singleton: EvalApplicative = object : arrow.core.extens
 @Deprecated(
   "@extension kinded projected functions are deprecated",
   ReplaceWith(
-  "just()",
-  "arrow.core.just"
+    "just()",
+    "arrow.core.just"
```
Comment (nomisRev) [inline arrow-core/src/main/kotlin/arrow/core/extensions/eval/applicative/EvalApplicative.kt:27]:
```suggestion
    "Eval.now(this)",
    "arrow.core.Eval"
```

File: arrow-core/src/main/kotlin/arrow/core/extensions/eval/applicative/EvalApplicative.kt [hunk 6]
```diff
@@ -51,14 +42,14 @@ fun <A> A.just(): Eval<A> = arrow.core.Eval.applicative().run {
 @Deprecated(
   "@extension kinded projected functions are deprecated",
   ReplaceWith(
-  "unit()",
-  "arrow.core.Eval.unit"
+    "just(Unit)",
+    "arrow.core.just"
```
Comment (nomisRev) [inline arrow-core/src/main/kotlin/arrow/core/extensions/eval/applicative/EvalApplicative.kt:46]:
```suggestion
    "Eval.Unit",
    "arrow.core.Eval"
```

File: arrow-core/src/main/kotlin/arrow/core/extensions/eval/apply/EvalApply.kt [hunk 7]
```diff
@@ -98,8 +93,8 @@ fun <A, B, Z> Kind<ForEval, A>.map2Eval(
 @Deprecated(
   "@extension kinded projected functions are deprecated",
   ReplaceWith(
-  "map(arg0, arg1, arg2)",
-  "arrow.core.Eval.map"
+    "Eval.mapN(arg0, arg1, arg2)",
+    "arrow.core.Eval"
```
Comment (nomisRev) [inline arrow-core/src/main/kotlin/arrow/core/extensions/eval/apply/EvalApply.kt:97]:
```suggestion
    "Eval.mapN(arg0, arg1) { a, b -> arg2(Tuple2(a, b)) }",
    "arrow.core.Eval", "arrow.core.Tuple2"
```

Idem for other `map` & `mapN` methods.

File: arrow-core/src/main/kotlin/arrow/core/extensions/eval/apply/EvalApply.kt [hunk 8]
```diff
@@ -612,15 +608,15 @@ fun <A, B, Z> Kind<ForEval, A>.map2(arg1: Kind<ForEval, B>, arg2: Function1<Tupl
 @Deprecated(
   "@extension kinded projected functions are deprecated",
   ReplaceWith(
-  "product(arg1)",
-  "arrow.core.product"
+    "Eval.mapN(this, fb) { a, b -> Tuple2(a, b) }",
+    "arrow.core.Eval"
```
Comment (nomisRev) [inline arrow-core/src/main/kotlin/arrow/core/extensions/eval/apply/EvalApply.kt:612]:
```suggestion
    "Eval.mapN(this, fb) { a, b -> Tuple2(a, b) }",
    "arrow.core.Eval", "arrow.core.Tuple2"
```

File: arrow-core/src/main/kotlin/arrow/core/extensions/eval/apply/EvalApply.kt [hunk 9]
```diff
@@ -632,15 +628,15 @@ fun <A, B> Kind<ForEval, A>.product(arg1: Kind<ForEval, B>): Eval<Tuple2<A, B>>
 @Deprecated(
   "@extension kinded projected functions are deprecated",
   ReplaceWith(
-  "product(arg1)",
-  "arrow.core.product"
+    "Eval.mapN(this, fb) { (ab, c) -> Tuple3(ab.a, ab.b, c) }",
+    "arrow.core.Eval"
```
Comment (nomisRev) [inline arrow-core/src/main/kotlin/arrow/core/extensions/eval/apply/EvalApply.kt:632]:
```suggestion
    "Eval.mapN(this, fb) { ab, c -> Tuple3(ab.a, ab.b, c) }",
    "arrow.core.Eval", "arrow.core.Tuple3"
```

Idem for other `product` methods

File: arrow-core/src/main/kotlin/arrow/core/extensions/eval/apply/EvalApply.kt [hunk 10]
```diff
@@ -800,15 +798,15 @@ fun <A, B, C, D, E, FF, G, H, I, Z> Kind<ForEval, Tuple9<A, B, C, D, E, FF, G, H
 @Deprecated(
   "@extension kinded projected functions are deprecated",
   ReplaceWith(
-  "tupled(arg0, arg1)",
-  "arrow.core.Eval.tupled"
+    "Eval.mapN(arg0, arg1) { a, b -> Tuple2(a, b) }",
+    "arrow.core.Eval"
```
Comment (nomisRev) [inline arrow-core/src/main/kotlin/arrow/core/extensions/eval/apply/EvalApply.kt:802]:
```suggestion
    "arrow.core.Eval", "arrow.core.Tuple2"
```

Same for other `tupled`/`tupledN` methods

File: arrow-core/src/main/kotlin/arrow/core/extensions/eval/comonad/EvalComonad.kt [hunk 11]
```diff
@@ -47,14 +42,15 @@ fun <A, B> Kind<ForEval, A>.coflatMap(arg1: Function1<Kind<ForEval, A>, B>): Eva
 @Deprecated(
   "@extension kinded projected functions are deprecated",
   ReplaceWith(
-  "extract()",
-  "arrow.core.extract"
+    "extract()",
+    "arrow.core.extract"
```
Comment (nomisRev) [inline arrow-core/src/main/kotlin/arrow/core/extensions/eval/comonad/EvalComonad.kt:46]:
```suggestion
    "value()"
```

File: arrow-core-data/src/main/kotlin/arrow/core/Eval.kt [hunk 12]
```diff
@@ -276,9 +422,37 @@ sealed class Eval<out A> : EvalOf<A> {
...
+
+  @JvmName("coflatMapKind")
+  @Deprecated("Kind is deprecated, and will be removed in 0.13.0. Please use the coflatMap method defined for Eval instead")
   inline fun <B> coflatMap(crossinline f: (EvalOf<A>) -> B): Eval<B> =
     Later { f(this) }
 
+  inline fun <B> coflatMap(crossinline f: (Eval<A>) -> B): Eval<B> =
+    Later { f(this) }
+
   fun extract(): A = value()
```
Comment (nomisRev) [inline arrow-core-data/src/main/kotlin/arrow/core/Eval.kt:456]: Deprecate in favor of `value()`. No need to duplicate the `Comonad` API here.

**Review by nomisRev — COMMENTED** [review]
A couple of incorrect deprecations, and nits.
Great work!! 👏 👏

**Review by nomisRev — APPROVED** [review]
Looks great @danimontoya! 👏 👏 👏 Thanks 🙏

**Comment by danimontoya** [comment]
@nomisRev thanks to you for all the reviews 🙌  😄

File: arrow-core-data/src/main/kotlin/arrow/core/Eval.kt [hunk 13]
```diff
@@ -242,6 +267,125 @@ sealed class Eval<out A> : EvalOf<A> {
...
+    ): Eval<C> =
+      mapN(a, b, Unit, Unit, Unit, Unit, Unit, Unit, Unit) { aa, bb, _, _, _, _, _, _, _ -> map(aa, bb) }
+
+    fun <A, B, C, D> mapN(
+      a: Eval<A>,
+      b: Eval<B>,
+      c: Eval<C>,
+      map: (A, B, C) -> D
+    ): Eval<D> =
+      mapN(a, b, c, Unit, Unit, Unit, Unit, Unit, Unit, Unit) { aa, bb, cc, _, _, _, _, _, _, _ -> map(aa, bb, cc) }
```
Comment (i-walker) [inline arrow-core-data/src/main/kotlin/arrow/core/Eval.kt:284]: nit:
with the kotlin naming conventions wouldn't we define `Unit` starting with a lowercase letter same there too
https://github.com/arrow-kt/arrow-core/blob/01781b8380401f0f1d2bc60edb234bcb23aec41b/arrow-core-data/src/main/kotlin/arrow/core/Eval.kt#L175-L183

**Review by i-walker — APPROVED** [review]
Looks great @danimontoya 🙌 
Just left a nit there, so those members are consistent across libraries in terms of naming conventions.

**Pull request merged**